[kafka] Update to kafka 0.9

brianwolfe opened the pull request:
## Basic version of kafka09 producer for Plog

This implements the very basic form of the kafka 09 producer in plog. It is the minimal form that lets us switch out the Kafka Partitioners in the new producer API. All performance improvements have been delayed for a later release. I wasn't able to get the unix domain sockets to perform as well as I originally anticipated for the use-case.

@luochenUmich

brianwolfe committed:
[kafka] Update to KafkaConsumer09

brianwolfe committed:
[kafka] Update to kafka version to 0.9

- Use new kafka producer
- Add a flink partitioner that uses murmur3 and key groups to partition
  in a way that is compatible with Flink.

brianwolfe committed:
[fixup] fix intending

Fix some weird indenting that I put in accidentally.

luochenUmich commented on plog-kafka/src/main/java/com/airbnb/plog/kafka/partitioner/FlinkPartitioner.java:
```diff
@@ -0,0 +1,106 @@
+
+  static int computePartition(Object key, int numPartitions, int maxParallelism) {
+    int group = murmurHash(key.hashCode()) % maxParallelism;
+    return (group * numPartitions) / maxParallelism;
```
Just to make sure I understand this correctly. Do we need to have the same number of consumerTasks and traceAggregationTasks as the number of kafka partitions to minimize shuffling?

brianwolfe replied:
No, it's slightly simpler than that. As long as `maxParallelism` matches, we can do a limited-fan-out ingestion job from this.

luochenUmich commented on plog-kafka/src/main/java/com/airbnb/plog/kafka/partitioner/FlinkPartitioner.java:
```diff
@@ -0,0 +1,106 @@
+  private int maxParallelism = 16386;
+
+  private static int toPositive(int number) {
+    return number & 2147483647;
```
nit: Change 2147483647 to `0x7fffffff ` for easier understanding.

luochenUmich reviewed (approved):
Looks good to me overall. Just some minor questions.

brianwolfe committed:
[flink-partitioner] Use Integer.MAX_VALUE

- Inline constant should instead be Integer.MAX_VALUE. That is always
  going to be the correct value and makes it more clear.
